ggu/main.c

```diff
@@ -16,11 +16,295 @@
+
+		ret = policy_parse_next(proto_ptr, policy_ptr, &policy);
+		if (ret < 0)
+			goto free_packet;
+
+		proto_ptr = proto_ptr + 1;
+		if (policy.flow.proto == ETHER_TYPE_IPv4)
+			policy_ptr = policy_ptr + sizeof(policy.flow.f.v4) +
+				sizeof(policy.params);
+		else
+			policy_ptr = policy_ptr + sizeof(policy.flow.f.v6) +
+				sizeof(policy.params);
+
+		process_single_policy(&policy, ggu_conf);
+	}
+
+free_packet:
+	rte_pktmbuf_free(pkt);
+}
+
+static int
+ggu_proc(void *arg)
+{
+	int ret;
+	uint32_t lcore = rte_lcore_id();
+	struct ggu_config *ggu_conf = (struct ggu_config *)arg;
+	uint8_t port_in = ggu_conf->net->back.id;
+	uint16_t rx_queue = ggu_conf->rx_queue_back;
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"ggu: the GK-GT unit is running at lcore = %u\n", lcore);
+
+	/* Wait for network devices to start. */
+	while (ggu_conf->net->configuring)
+		;
+
+	/*
+	 * Setup the ntuple filter that assign the GK-GT packets
+	 * to its queue for both IPv4 and IPv6 addresses.
+	 * The ntuple Filters should be configured
+	 * after the network devices are started.
+	 */
+	ret = ntuple_filter_add(port_in,
+		ggu_conf->net->back.ip4_addr.s_addr,
+		rte_cpu_to_be_16(ggu_conf->ggu_src_port),
+		rte_cpu_to_be_16(ggu_conf->ggu_dst_port),
+		ggu_conf->rx_queue_back);
+	if (ret < 0) {
+		exiting = true;
+		goto out;
+	}
+
+	while (likely(!exiting)) {
+		uint16_t i;
+		uint16_t num_rx;
+		struct rte_mbuf *bufs[GATEKEEPER_MAX_PKT_BURST];
+
+		/* Load a set of GK-GT packets from the back NIC. */
+		num_rx = rte_eth_rx_burst(port_in, rx_queue, bufs,
+			GATEKEEPER_MAX_PKT_BURST);
+		if (unlikely(num_rx == 0))
+			continue;
+
+		for (i = 0; i < num_rx; i++)
+			process_single_packet(bufs[i], ggu_conf);
+	}
+
+out:
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"ggu: the GK-GT unit at lcore = %u is exiting\n", lcore);
+	return cleanup_ggu(ggu_conf);
+}
 
 int
-run_ggu(__attribute__((unused)) const struct ggu_config *ggu_conf)
+run_ggu(struct net_config *net_conf,
+	struct gk_config *gk_conf, struct ggu_config *ggu_conf)
 {
-	/* TODO Initialize and run GK-GT Unit functional block. */
+	int ret;
+
+	if (ggu_conf == NULL) {
+		ret = -1;
+		goto out;
+	}
+
+	if (net_conf == NULL || gk_conf == NULL) {
+		ret = -1;
+		goto configuration;
+	}
+
+	ggu_conf->net = net_conf;
+	ggu_conf->gk = gk_conf;
+
+	ret = get_queue_id(&ggu_conf->net->back,
+		QUEUE_TYPE_RX, ggu_conf->lcore_id);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER, "gk: cannot assign an RX queue for the back interface for lcore %u\n",
+			ggu_conf->lcore_id);
+		goto configuration;
+	}
+	ggu_conf->rx_queue_back = (uint16_t)ret;
+
+	ret = rte_eal_remote_launch(ggu_proc, ggu_conf, ggu_conf->lcore_id);
+	if (ret) {
+		RTE_LOG(ERR, EAL, "lcore %u failed to launch GGU\n",
+			ggu_conf->lcore_id);
+		ret = -1;
+		goto configuration;
+	}
+
+	ret = 0;
+	goto out;
+
+configuration:
+	/*
+	 * XXX Review this deallocation.
+	 * Should it be done here, in Lua, or elsewhere?
+	 */
+	rte_free(ggu_conf);
+	ggu_conf = NULL;
```

Comment:
Do you mean we need to free ``ggu_conf`` in Lua script? I think both variables (i.e., one in Lua, and one in C) will point to the same memory location, as long as we set the variable in Lua as ``nil``, that should be fine? Since even the program runs correctly, we also need to close the Lua script after setting up the program, and leaves a pointer to the ``ggu_conf`` in C.